File: evaluation/chord_matcher.py
from __future__ import print_function
import yaml
import re
import sys

class ChordMatcher(object):
    def __init__(self, config):
        self._load_config(config)

    def _load_config(self, filename):
        config_file = open(filename)
        config = yaml.load(config_file)
        config_file.close()

        self.pitches = config["pitches"]
        self.shorthands = config["shorthands"]
        self.intervals = config["intervals"]
        self.pitch_names = config["pitch_names"]

    def chords_match(self, ca, cb):
        return self.chord_notes(ca) == self.chord_notes(cb)

    def chord_notes(self, chord):
        pitch = chord.split(":")[0]
        notes = [self.parse_pitch(pitch)] + self.parse_shorthand(chord)
        intervals = self.get_intervals(chord)

        if chord.count("/") == 1:
            intervals.append(chord.split("/")[-1])

        for interval in intervals:
            if len(interval) >= 1:
                if interval[0] == "*":
                    notes.remove(self.parse_interval(interval[1:]))
                else:
                    notes.append(self.parse_interval(interval))

        if len(notes) == 1:
            notes += self.shorthands['maj']

        return set(notes)

    def get_intervals(self, chord):
        found = re.findall("(?<=\().*(?=\))", chord)
        if len(found) > 0:
            return found[0].split(",")
        return []

    def parse_shorthand(self, chord):
        found = re.findall("(?<=:)[^()/]+(?=\(|\/|$)", chord)
        if len(found) == 0:
            return []
        if found[0] in self.shorthands.keys():
            return self.shorthands[found[0]]
        else:
            print("ERROR! Invalid shorthand: %s" % found[0])
            return []

    def parse_interval(self, interval_str):
        i = self.intervals[str(int(re.split("b|#", interval_str)[-1]) % 8)]
        return i - interval_str.count("b") + interval_str.count("#")

    def parse_pitch(self, pitch):
        i = self.pitches[re.split("b|#", pitch)[0]]
        return self.pitch_names[i - pitch.count("b") + pitch.count("#")]

File: src/chord_matcher.rb
class ChordMatcher
  def initialize(config_file=nil, debug: false)
    @debug = debug

    config_file ||= File.expand_path("../chord_matcher.yml", __FILE__)
    config = YAML.load_file(config_file)
    @shorthands = config["shorthands"]
    @intervals = config["intervals"]

    pitches = %w(C C# D D# E F F# G G# A A# B)
    @pitch_index = pitches.map.with_index.to_h

    @main_chord_templates = {}
    pitches.each do |pitch|
      %w(maj min dim aug).each do |shorthand|
        label = [pitch, shorthand != "maj" ? shorthand : nil].compact.join(":")
        @main_chord_templates[notes(label)] = label
      end
    end
  end

  def normalize(chord)
    @main_chord_templates[notes(chord)]
  end

  def compare(chord_a, chord_b)
    print "#{chord_a} \t #{chord_b}\n" if @debug
    return false if chord_a == "N" || chord_b == "N"
    [notes(chord_a), notes(chord_b)].transpose.map { |a, b| a <= b }.all?
  end

  def notes(chord)
    return [nil, []] if chord == "N"

    notes = Array.new(12, 0)

    root = parse_pitch(chord)
    added_notes, removed_notes = parse_intervals(root, chord)
    shorthand_notes = parse_shorthand(root, chord)

    if [added_notes + shorthand_notes].flatten.size == 0
      added_notes = parse_shorthand(root, ":maj")
    end

    indices_to_add = [
      root, shorthand_notes,
      parse_bass(root, chord),
      added_notes
    ].flatten.uniq

    indices_to_add.each { |i| notes[i] = 1 }
    removed_notes.each { |i| notes[i] = 0 }

    notes
  end

  def parse_pitch(chord)
    pitch = chord.split(":")[0]
    pitch = pitch.split("/")[0]
    root = pitch.delete("b#")
    (@pitch_index[root] + pitch.count("#") - pitch.count("b")) % 12
  end

  def parse_shorthand(root, chord)
    shorthand = chord[/(?<=:)[^()\/]+/]
    notes = @shorthands[shorthand]
    p "ERROR! Invalid shorthand: #{shorthand}" if shorthand && notes.nil?
    sum_root(root, notes.to_a)
  end

  def parse_intervals(root, chord)
    added_notes = []
    removed_notes = []

    intervals = chord[/(?<=\().*(?=\))/].to_s

    intervals.split(",").map { |s| s.strip }.each do |interval|
      if interval[0] == "*"
        removed_notes << parse_interval(interval[1..-1])
      else
        added_notes << parse_interval(interval)
      end
    end

    [sum_root(root, added_notes), sum_root(root, removed_notes)]
  end

  def parse_interval(interv)
    i = interv[/\d+/].to_i
    i %= 7 if i > 7
    (@intervals[i] + interv.count("#") - interv.count("b")) % 12
  end

  def parse_bass(root, chord)
    bass = chord[/(?<=\/).+$/]
    bass ? sum_root(root, [parse_interval(bass)]) : []
  end

  def sum_root(root, notes)
    notes.map { |note| (note + root) % 12 }
  end
end
